river/cluster/evostream.py

```diff
@@ -0,0 +1,424 @@
+    inspired by natural evolution where promising solutions are combined to create
+    offsprings which can combine the best arrtributes of both parents.
+
+    In River, since data arrives continuously, we will replace the concept of "idle" time
+    with an evolution gap. This means that, after each gap, a temporary batch will be
+    formed, the evolution function will be called, simulating the effect of "idle"
+    time between batches of data points.
+
+    The algorithm is divided into two main parts:
+
+    **Online micro-cluster maintenance (learning)**
+
+    For a new point `p`:
+
+    * Create a new micro cluster from point `p`. Find all micro clusters whose
+    centers fall within the fixed radius (clustering threshold) from the newly
+    generated cluster
+
+    * If one or more neighboring micro clusters are found, these micro clusters
+    will be updated, trying to move them closer to the new instance (similar to the
+    DBSTREAM algorithm [^3]). Else, that new micro cluster will
+    be added to the current.
+
+    * After each cleanup interval, weights of all micro clusters are updated once
+    again. At this time, if the weight of any cluster is less than a certain threshold,
+    it will be removed. Then, all clusters whose centers are less than `r` from
+    each other are merged.
+
+    * When the stream is not initiated and the number of cluster is at least equal
+    to the initialization threshold, a cluster population of size `P` will be generated
+    by choosing randomly micro-clusters from the set of currently available micro-clusters.
+
+    **Offline generation of macro clusters (evolutionary step)**
+
+    The offline generation of macro clusters is based upon the original GA clustering algirthm,
+    as follows:
+
+    * Using the roulette wheel algorithm, choose 2 arbitrary cluster solutions proportionally
+    to their fitness from the cluster population.
+
+    * Generate two offsprings using binary crossover.
+
+    * Generate a random number `delta` within the range `[0,1]`. If `delta < P_m`, each gene `g_i`
+    within these two offsprings will be mutated with the following rules
+
+        - If `g_i = 0`, `g_i` will be equal to either `-2 * delta` or `2 * delta`, with equal probability
+
+        - If `g_i != 0`, `2 * delta * g_i` will be added or subtracted from `g_i`, with equal probability
+
+    * The two newly generated solutions will be added to the set of cluster population. Two solutions with
+    highest SSQ (or lowest fitness) will be discarded.
+
+    * Choose the solution with the highest fitness and assign it as the clusters for the whole algorithm.
+
+    Parameters
+    ----------
+    radius
+        Radius threshold to determine neighboring clusters.
+    decay_rate
+        Parameter that controls the importance of historical data to current cluster.
+        Note that `decay_rate` is usually sufficiently small and has to be different from `0`.
```

Comment:
Suggestion: "Note that `decay_rate` should be sufficiently small and greater than zero."

Reply:
Suggestion added!